components/web_discovery/browser/server_config_loader.cc

```diff
@@ -0,0 +1,444 @@
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/server_config_loader.h"
+
+#include <utility>
+
+#include "base/barrier_callback.h"
+#include "base/base64.h"
+#include "base/containers/fixed_flat_set.h"
+#include "base/files/file_util.h"
+#include "base/functional/bind.h"
+#include "base/json/json_reader.h"
+#include "base/location.h"
+#include "base/rand_util.h"
+#include "base/task/thread_pool.h"
+#include "base/values.h"
+#include "brave/components/web_discovery/browser/pref_names.h"
+#include "brave/components/web_discovery/browser/util.h"
+#include "components/prefs/pref_service.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "third_party/zlib/google/compression_utils.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr base::TimeDelta kMinReloadInterval = base::Hours(1);
+constexpr base::TimeDelta kMaxReloadInterval = base::Hours(4);
+
+constexpr base::TimeDelta kPatternsMaxAge = base::Hours(2);
+constexpr base::TimeDelta kPatternsRequestLatestDelay = base::Minutes(3);
+constexpr base::TimeDelta kPatternsRequestInitDelay = base::Seconds(15);
+
+constexpr size_t kPatternsMaxFileSize = 128000;
+
+constexpr net::NetworkTrafficAnnotationTag kNetworkTrafficAnnotation =
+    net::DefineNetworkTrafficAnnotation("wdp_config", R"(
+    semantics {
+      sender: "Brave Web Discovery Server Configuration Fetch"
+      description:
+        "Requests server configuration needed to send Web Discovery "
+        "measurements to Brave servers."
+      trigger:
+        "Requests are automatically sent at intervals "
+        "while Brave is running."
+      data: "Configuration attributes"
+      destination: WEBSITE
+    }
+    policy {
+      cookies_allowed: NO
+      setting:
+        "Users can opt-in or out via brave://settings/search"
+    })");
+
+constexpr char kGroupPubKeysFieldName[] = "groupPubKeys";
+constexpr char kPubKeysFieldName[] = "pubKeys";
+constexpr char kMinVersionFieldName[] = "minVersion";
+constexpr char kKeysFieldName[] = "keys";
+constexpr char kLimitFieldName[] = "limit";
+constexpr char kPeriodFieldName[] = "period";
+constexpr char kSourceMapFieldName[] = "sourceMap";
+constexpr char kSourceMapActionsFieldName[] = "actions";
+constexpr char kLocationFieldName[] = "location";
+
+constexpr char kCollectorConfigPathWithFields[] =
+    "/config?fields=minVersion,groupPubKeys,pubKeys,sourceMap";
+constexpr char kQuorumConfigPath[] = "/config";
+constexpr char kPatternsFilename[] = "wdp_patterns.json";
+
+constexpr char kOmittedLocationValue[] = "--";
+constexpr auto kAllowedReportLocations =
+    base::MakeFixedFlatSet<std::string_view>(
+        {"de", "at", "ch", "es", "us", "fr", "nl", "gb", "it", "be",
```

Comment:
nit: sorting this list would make it easier to read

Reply:
072fdfc3176b60fd71cedf9c05ed83a3420e7415